pkg/utils/bpf-iter-ns/bpf-iter-ns.go

```diff
@@ -103,12 +120,30 @@ func ReadOnHostPidNs(iter *link.Iter) ([]byte, error) {
 	var wg sync.WaitGroup
 	wg.Add(1)
 	go func() {
+		defer wg.Done()
+		defer r.Close()
+
+		var b bytes.Buffer
+		readBuf := make([]byte, 8192)
 		stdoutReader := bufio.NewReader(r)
-		// ReadAll will block until the write-side of the pipe is closed in both processes
-		// (the systemd service and this process)
-		buf, errReader = io.ReadAll(stdoutReader)
-		r.Close()
-		wg.Done()
+
+		for {
+			n, err := stdoutReader.Read(readBuf)
+			if n > 0 {
+				b.Write(readBuf[:n])
+			}
+			if err == io.EOF {
+				break
+			}
+			if errors.Is(err, syscall.EAGAIN) {
+				continue
+			}
```

Comment:
Here it will not work: `stdoutReader.Read(readBuf)` is reading from a pipe fd and not from the bpf iter fd.
The bpf iter is actually read by a different process, "cat", see `fmt.Sprintf("cat %s > %s", pinPathFromHost, writerPath)` nelow. We could replace "cat" by something else that ignores EAGAIN, such as socat if available (to be tested):
```go
fmt.Sprintf("{ command -v socat &> /dev/null && socat -u STDIN STDOUT || cat; } < %s > %s", pinPathFromHost, writerPath),
```

For the socket enricher, the patch from #4942 should be enough to fix the bug. Your patch on ReadOnCurrentPidNs above is still good to take because this function is not only used by the socket enricher but also by snapshot gadgets (snapshot_process and snapshot_socket).